File: C#/KataClasses/DigitalRoot.cs
//Codewars

/*===============================================================================================================
    Given n, take the sum of the digits of n.
    If that value has more than one digit, continue reducing in this way until a single-digit number is produced.
    The input will be a non-negative integer.
===============================================================================================================*/

namespace Katas.KataClasses;

public static class DigitalRoot
{
    public static int ReturnDigitalRoot(long n)
    {
        var num = n.ToString().ToCharArray().Select(e => Int32.Parse(e.ToString())).Aggregate((a, b) => a + b);

        if (num.ToString().Length > 1)
            return ReturnDigitalRoot(num);

        return num;
    }
}

/*
    Notes:

    I forgot about .Sum() which would have have worked just as well as Aggregate. It would have also
    been more readable/self-documenting for someone else.
 */

File: TS_JS/TypeScript/KataClasses/digitalRoot.ts
// Codewars

/*==========================================================================================================
Given n, take the sum of the digits of n.
If that value has more than one digit, continue reducing in this way until a single-digit number is produced.
The input will be a non-negative integer.
==========================================================================================================*/

export const digitalRoot = (n: number): number => {
    let numArr = n;

    do {
        numArr = numArr
            .toString()
            .split("")
            .map((e) => parseInt(e))
            .reduce((a, b) => a + b, 0);
    } while (numArr.toString().length > 1);

    return numArr;
};

//Surely the Electric Boogaloo Joke will never get old.
export const digitalRoot2ElectricBoogaloo = (n: number): number => {
    let num = n
        .toString()
        .split("")
        .map((e) => parseInt(e))
        .reduce((a, b) => a + b, 0);

    if (num.toString().length > 1) {
        return digitalRoot2ElectricBoogaloo(num);
    }
    
    return num;
};
